nccl_overlap/jacobi.cpp

```diff
@@ -166,19 +169,25 @@ int main(int argc, char* argv[]) {
     const int ny = get_argval<int>(argv, argv + argc, "-ny", 16384);
     const bool csv = get_arg(argv, argv + argc, "-csv");
 
-    int local_rank = -1;
-    {
+    if(num_devices > 1) {
+        int local_rank = -1;
+     
         MPI_Comm local_comm;
         MPI_CALL(MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_SHARED, rank, MPI_INFO_NULL,
                                      &local_comm));
 
         MPI_CALL(MPI_Comm_rank(local_comm, &local_rank));
 
         MPI_CALL(MPI_Comm_free(&local_comm));
+
+        CUDA_RT_CALL(cudaSetDevice(local_rank));
+        CUDA_RT_CALL(cudaFree(0));
     }
 
-    CUDA_RT_CALL(cudaSetDevice(local_rank));
-    CUDA_RT_CALL(cudaFree(0));
+    else {
+        CUDA_RT_CALL(cudaSetDevice(0));
+        CUDA_RT_CALL(cudaFree(0));
```

Comment:
Please move `CUDA_RT_CALL(cudaFree(0));` to after the if/else block as it is executed in both cases.